File: datastruct/go/02_链表/list.go
/*
	简单链表
*/
package main

import (
	"fmt"
)

// 节点结构
type Node struct {
	data int
	next *Node
}

// 链表结构
type List struct {
	length int
	head *Node
}

// 初始化
func New() *List {
	return &List{}
}

// 判断是否为空
func (list *List) isEmpty() bool {
	return list.head == nil
}

// 追加
func (list *List) append(e int) {
	if list.head == nil {
		list.head = &Node{data: e}
		list.length++
		return
	}
	head := list.head
	for head.next != nil {
		head = head.next
	}
	head.next = &Node{data: e}
	list.length++
}

// 获取链表长度
func (list *List) getSize() int {
	return list.length
}

// 反转链表
func (list *List) reverse() {
	var p *Node
	head := list.head
	for head != nil {
		q := head
		head = q.next
		q.next = p
		p = q
	}
	list.head = p
}

// 遍历所有节点
func (list *List) printAll() {
	idx := 0
	head := list.head
	for head != nil {
		idx++
		fmt.Printf("idx: %d, data: %d\n", idx, head.data)
		head = head.next
	}
}

func main() {
	fmt.Println("----简单链表----")
	list := New()
	list.append(3)
	list.append(5)
	fmt.Println("链表长度:", list.getSize())
	fmt.Println("链表是否为空:", list.isEmpty())
	list.printAll()
	list.reverse()
	list.printAll()
}

File: datastruct/js/01_链表/list.js
/*
    链表
*/

class Node {
  constructor(data, next = null) {
    this.data = data
    this.next = next
  }
}

class List {
  constructor() {
    // 初始化
    this.head = null
    this.length = 0
  }

  isEmpty() {
    return this.length == 0
  }

  getSize() {
    return this.length
  }

  append(e) {
    /*
      从head开始以为循环到尾部, 追加新的节点
      如果创建pop函数也是一样
    */
    if (this.head === null) {
      this.head = new Node(e)
      this.length += 1
    } else {
      let head = this.head
      while (head.next !== null) {
        head = head.next
      }
      head.next = new Node(e)
      this.length += 1
    }
  }

  reverse() {
    /*
      反转链表
      移动head指针，指向下个节点， 并把原来节点指向上一个节点
    */
    let p = null
    let head = this.head
    while (head !== null) {
      // 把head赋值给q, 这里q和head都指向同一个Node
      let q = head
      // 然后head指向给了下一个节点
      head = q.next
      // q指向上一个节点
      q.next = p
      // 把当前节点指向为上一个节点
      p = q
    }
    this.head = p
  }

  print() {
    let idx = 0
    let head = this.head
    while (head !== null) {
      idx += 1
      console.log(`idx ${idx}, data: ${head.data}`)
      head = head.next
    }
  }
}

;(function main() {
  let list = new List()
  list.append(3)
  list.append(5)
  list.append(1)
  list.print()
  list.reverse()
  list.print()
})()
